src/ir/si_builder.cc

```diff
@@ -0,0 +1,341 @@
+  }
+  StmtExprVisitor::VisitStmt(stmt);
+}
+
+class TirRecursivelyFill : public tir::StmtExprMutator {
+ public:
+  TirRecursivelyFill(const Span& span, const PrimExprSet& expr_inputs = {},
+                     const StmtSet& stmt_inputs = {})
+      : span_(span), expr_inputs_(expr_inputs), stmt_inputs_(stmt_inputs) {}
+
+  tir::Stmt Fill(const tir::Stmt& s) { return operator()(s); }
+  PrimExpr Fill(const PrimExpr& e) { return operator()(e); }
+
+  bool IsInput(const PrimExpr& expr);
+  bool IsInput(const tir::Stmt& stmt);
+
+  PrimExpr VisitExpr(const PrimExpr& expr) final;
+  tir::Stmt VisitStmt(const tir::Stmt& stmt) final;
+
+ private:
+  const Span& span_;
+  const PrimExprSet& expr_inputs_;
+  const StmtSet& stmt_inputs_;
+};
+
+tir::Stmt TirRecursivelyFill::VisitStmt(const tir::Stmt& stmt) {
+  if (IsInput(stmt)) {
+    return stmt;
+  }
+  stmt->span = span_;
+  return StmtExprMutator::VisitStmt(stmt);
+}
+
+bool TirRecursivelyFill::IsInput(const PrimExpr& expr) {
+  return expr_inputs_.find(expr) != expr_inputs_.end();
+}
+
+bool TirRecursivelyFill::IsInput(const tir::Stmt& stmt) {
+  return stmt_inputs_.find(stmt) != stmt_inputs_.end();
+}
+
+PrimExpr TirRecursivelyFill::VisitExpr(const PrimExpr& expr) {
+  if (IsInput(expr)) {
+    return expr;
+  }
+  expr->span = span_;
+  return StmtExprMutator::VisitExpr(expr);
+}
+
+struct SIBuilder::Impl {
+  virtual Span CreateSpan() const = 0;
+  virtual void RecursivelyFillSpan(const relay::Expr& entry, const RelayExprSet& inputs) const = 0;
+  virtual void RecursivelyFillSpan(const PrimExpr& entry, const PrimExprSet& inputs) const = 0;
+  virtual void RecursivelyFillSpan(const tir::Stmt& entry, const PrimExprSet& inputs) const = 0;
+  virtual void RecursivelyFillSpan(const tir::Stmt& entry, const StmtSet& inputs) const = 0;
+  virtual void CollapseSpan(const relay::Expr& entry, const RelayExprSet& inputs) = 0;
+  virtual void CollapseSpan(const PrimExpr& entry, const PrimExprSet& inputs) = 0;
+  virtual void CollapseSpan(const tir::Stmt& entry, const PrimExprSet& inputs) = 0;
+  virtual void CollapseSpan(const tir::Stmt& entry, const StmtSet& inputs) = 0;
+};
+
+SIBuilder::~SIBuilder() = default;
+
+Span SIBuilder::CreateSpan() const { return impl_->CreateSpan(); }
+
+template <>
+void SIBuilder::RecursivelyFillSpan(const relay::Expr& entry, const RelayExprSet& inputs) const {
+  impl_->RecursivelyFillSpan(entry, inputs);
+}
+
+template <>
+void SIBuilder::RecursivelyFillSpan(const PrimExpr& entry, const PrimExprSet& inputs) const {
+  impl_->RecursivelyFillSpan(entry, inputs);
+}
+
+void SIBuilder::RecursivelyFillSpan(const tir::Stmt& entry, const PrimExprSet& inputs) const {
+  impl_->RecursivelyFillSpan(entry, inputs);
+}
+
+void SIBuilder::RecursivelyFillSpan(const tir::Stmt& entry, const StmtSet& inputs) const {
+  impl_->RecursivelyFillSpan(entry, inputs);
+}
+
+std::unique_ptr<SIBuilder::Impl> SIBuilder::CreateImpl(const Span& span) {
+  struct NullImpl : public SIBuilder::Impl {
```

Comment:
This class and the base class `SIBuilder::Impl ` don't seem to be necessary. If SI collection is not enabled, there is no need to collect spans or create the impl object. We should check `enable_si_builder` condition much earlier.

Reply:
What I meant was, if si collection is not enabled, we don't have to call `CreateImpl` at all. What you described still sounds too complicated for its purpose.

Reply:
I see. I agree this implementation seems to be a bit too complicated in this case.
Basically we had two strategies to enable/disable the functionalities of SIBuilder. May you
tell us which one you prefer more or please kindly give us some advices? :D

In the following, I take a real case from one of our upcoming pass transformation PR for example,
I will list pro and con of our two strategies for you.

# The use case of SIBuilder
Here we uses `SimplifyReshape` in `${tvm}/src/relay/transforms/simplify_expr.cc` for example. It is
a many-to-one transformation. The SIBuilder will be invoked like following block:

```C++
class SimplifyReshape : public DFPatternRewrite {
 public:
  SimplifyReshape() {
    x_ = IsWildcard();
    auto reshape1 = IsOp("reshape") || IsOp("contrib_reverse_reshape");
    auto reshape2 = IsOp("reshape") || IsOp("contrib_reverse_reshape");
    pattern_ = reshape1({reshape2({x_})});
  }

  Expr Callback(const Expr& pre, const Expr& post,
                const Map<DFPattern, Array<Expr>>& node_map) const override {
    //...
    if (const_shape) {
      auto x = node_map[x_][0];
      auto ret = MakeReshape(x, newshape);

      SIBuilder si_builder(/* entry */ node_map[pattern_][0], /* inputs */ {x});
      ret->span = si_builder.Build();

      return ret;
    }
    return post;
  }
//..
};
```

In above case the si_builder is initialized by a sub graph, and invokes its member function to
build a new span. To make its functionalities enabled/disabled, we have 2 proposals

# Proposals
1. Null implementation
Just like what we have now. The pro is, we check `enable_si_builder` in a place(CreateImpl) only.
The con is obviously, it seems to be complicated in this case.

2. Embed `enable_si_builder` checking in each place
Another way is we can embed `enable_si_builder` checking at the beginning constructors or functions.
Take RecursivelyFillSpan an example, we can add the checking like this:

```C++
template <>
void SIBuilder::RecursivelyFillSpan(const relay::Expr& entry, const RelayExprSet& inputs) const {
    if (enabled) {
        RecursivelyFillSpan(entry, inputs);
    }
}
```
The pro is, the codes are very straightforward now. The con is the duplicated checking of
`enable_si_builder` will be added to multiple constructors and functions. It might seem to be a bit
redundant.

Currently we use the first proposal because we can reduce the checking to a place only, though it
is a bit complicated. Yet we are good to use the proposal 2. Or perhaps may you kindly point out
some strategies we might miss out?

Thank you for reading this long rely. We will get back to you soon. :D

Reply:
Yes, 2. is what I meant by "checking early". If that ends up with too-much repetitive check, I'm ok with your solution of making `NullImpl` a base class.

Reply:
Thank you very much for clarifying and understanding. :D
We would like to keep this NullImpl. Because we made many constructors to serve both Relay and TIR, the second proposal will lead us to add quite a lot number of checking (perhaps more than ten). 

We will fix the the codes based your comments and push it once we finished.